### PR TITLE
cm-tarnation grammar 2.0, performance, many other improvements

### PR DESCRIPTION
**Pull request by Monkatraz**
Highlights:
- Entirely new grammar format, written in JSON/YAML. Fairly readable and easy to understand, very flexible.
- Much better written grammar compiler and dumb-tokenizer
- Big performance improvements, much better right-hand reuse of previous tokenizer data
- Parsing of non-contiguous ranges works now
- Lots of bugfixes
- Trimmed a bunch of fat from the module in general
- Improvements to how much memory cached data uses

There are probably (quite) a few more bugs to fix, but they're nothing that's going to require any significant changes to the current code. Considering the quantity of code, I figure it's worth getting this PR open now. The main things to look at are `ftml.yaml` and the new grammar stuff.

### CODE REVIEW

File: client/modules/cm-lang-ftml/src/grammars/print-tree.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+repository:
+  Punctuation:
+    match: /[\u251C\u2514\u2502\u2500]/
```
Comment (emmiegit) [inline client/modules/cm-lang-ftml/src/grammars/print-tree.yaml:3]: What are these characters? Could you leave a comment explaining it?

File: client/modules/cm-lang-ftml/src/grammars/tex.yaml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  - Comment
+  - Function
+  - CommandGroup
+  - Command
+  - Spacing
+  - EscapedLine
+  - Special
+  - Number
+  - Operator
+  - Bracket
```
Comment (emmiegit) [inline client/modules/cm-lang-ftml/src/grammars/tex.yaml:63]: Why is this file here?

File: client/modules/cm-lang-ftml/src/grammars/tokens.yaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    captures:
+      0: { open: TokenPosition, tag: squareBracket }
+      1: { type: TokenFrom, tag: integer }
+      2: { type: TokenPositionArrow, tag: operator }
+      3: { type: TokenTo, tag: integer }
+      4: { close: TokenPosition, tag: squareBracket }
+      5: { type: TokenColon, tag: separator }
+      # 6:
+      7: { type: TokenName, tag: content }
+      # 8:
```
Comment (emmiegit) [inline client/modules/cm-lang-ftml/src/grammars/tokens.yaml:13]: Couldn't you just make these non-capturing groups in the regex?

File: client/modules/cm-tarnation/src/grammar/node.ts [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+ * (!) tag
+ * (!) func(tag)
+ * (...) tag
+ * (...) func(tag)
+ * (parent/) tag
+ * (parent/) func(tag)
+ * (grandparent/parent) tag
+ * (grandparent/parent) func(tag)
+ * ```
+ */
```
Comment (emmiegit) [inline client/modules/cm-tarnation/src/grammar/node.ts:94]: I don't think I understand, what is a "tag" in this context?

File: client/web/dev-sandbox/public/static/misc/ftml-test-cases-export.ftml [hunk 5]
```diff
@@ -15,11 +15,6 @@ Durian
 [[/>]]
 [[/=]]
 
-[!-- align-fail --]
```
Comment (emmiegit) [inline client/web/dev-sandbox/public/static/misc/ftml-test-cases-export.ftml:18]: Is this / should this be marked as generated?

File: client/modules/cm-lang-ftml/src/grammars/ftml.yaml [hunk 6]
```diff
@@ -0,0 +1,999 @@
...
+      1: { type: IncludeVariableName, tag: variableName }
+      2: { close: IncludeVariable, tag: brace }
+
+  PageVariable:
+    match: /(%%)(.*?)(\{.*?\})?(%%)/
+    captures:
+      0: { open: PageVariable, tag: bracket }
+      1: { type: PageVariableName, tag: variableName }
+      2: { type: PageVariableAccessor, tag: string }
+      3: { close: PageVariable, tag: bracket }
```
Comment (emmiegit) [inline client/modules/cm-lang-ftml/src/grammars/ftml.yaml:176]: gah I hate these lol

File: client/modules/cm-lang-ftml/src/grammars/ftml.yaml [hunk 7]
```diff
@@ -0,0 +1,999 @@
...
+      # center
+      - match: /(@cs)(=@ws+)/
+        captures:
+          0: { is: ContainerMark }
+          1: { is: CenterMark }
+      # everything else
+      - ContainerMark
+      - include: inline
+      - include: blocks
+
```
Comment (emmiegit) [inline client/modules/cm-lang-ftml/src/grammars/ftml.yaml:315]: For the record, what's going wrong with paragraphs? (I know they're a pain, I'm just curious)

File: client/modules/cm-lang-ftml/src/grammars/ftml.yaml [hunk 8]
```diff
@@ -0,0 +1,999 @@
...
+    match: /\]/
+    tag: squareBracket
+    openedBy: SingleLinkStart
+
+  SingleLinkSeparator:
+    match: "|"
+    tag: separator
+
+  SingleLinkSlug:
+    match: /([:#*]|(?=\/)|(?=[^#*\s]+?[@:][^#*\s]))([^#*\s]+)/
```
Comment (emmiegit) [inline client/modules/cm-lang-ftml/src/grammars/ftml.yaml:384]: I think this one is complicated enough to deserve a comment.

File: client/modules/cm-lang-ftml/src/grammars/ftml.yaml [hunk 9]
```diff
@@ -0,0 +1,999 @@
...
+        - BlockNameSuffix?
+        - BlockEnd
+
+  BlockNameCSS:
+    match: css
+    lookahead: /@namela/
+    emit: BlockName
+    tag: special(tagName)
+
+  BlockNestedCSS:
```
Comment (emmiegit) [inline client/modules/cm-lang-ftml/src/grammars/ftml.yaml:613]: Why is CSS allowed to be nested?

File: client/modules/cm-lang-ftml/src/grammars/ftml.yaml [hunk 10]
```diff
@@ -0,0 +1,999 @@
...
+  - EscapedNewLine
+  - EscapedCharacter
+  - BlockComment
+  - NonControl
+
+root:
+  - include: block
+  - include: includes
+  - include: blocks
+  - include: inline
```
Comment (emmiegit) [inline client/modules/cm-lang-ftml/src/grammars/ftml.yaml:999]: I'll say, despite the length, this YAML format is way cleaner to read than the previous version. +1

**Review by emmiegit — APPROVED** [review]

File: client/modules/cm-lang-ftml/src/grammars/print-tree.yaml (hunk 1 above)
Comment (Monkatraz) [reply client/modules/cm-lang-ftml/src/grammars/print-tree.yaml:3]: Oh they're the pipe indent marking characters when printing the tree. I should just make this a lookup, rather than a regex.

File: client/modules/cm-lang-ftml/src/grammars/tex.yaml (hunk 2 above)
Comment (Monkatraz) [reply client/modules/cm-lang-ftml/src/grammars/tex.yaml:63]: It's a simplistic grammar for wiki-math blocks, so it's a grammar for TeX. It is actually hooked up.

Comment (emmiegit) [reply client/modules/cm-lang-ftml/src/grammars/tex.yaml:63]: For inside those blocks? Incredible.

File: client/modules/cm-lang-ftml/src/grammars/tokens.yaml (hunk 3 above)
Comment (Monkatraz) [reply client/modules/cm-lang-ftml/src/grammars/tokens.yaml:13]: I need to properly document this, but JS regexes require a performance hit (and a modern browser) to get the indices for each capture group. So, the way Tarnation works is that if a regexp has capturing groups, the capturing groups must add up to the entire match with no gaps. That's why I skip these capture groups, they're just there to satisfy that rule.

Comment (emmiegit) [reply client/modules/cm-lang-ftml/src/grammars/tokens.yaml:13]: Ah okay. Yes please document this somewhere.

File: client/modules/cm-tarnation/src/grammar/node.ts (hunk 4 above)
Comment (Monkatraz) [reply client/modules/cm-tarnation/src/grammar/node.ts:94]: It's a CodeMirror highlighting tag. CodeMirror highlights stuff based on what tag it has, which basically helps with abstracting away theming. e.g. strings in any language will probably be marked as `string`. There are modifiers for these tags, such as wrapping special tags (e.g. `function(name)`), or only highlighting with a certain tag if the target node has a specific parent.

Basically, I made a syntax to let you get your grubby hands on all of these modifiers and the like.

Comment (emmiegit) [reply client/modules/cm-tarnation/src/grammar/node.ts:94]: Huh, okay. Could you enhance this comment to mention that it's CodeMirror?

File: client/web/dev-sandbox/public/static/misc/ftml-test-cases-export.ftml (hunk 5 above)
Comment (Monkatraz) [reply client/web/dev-sandbox/public/static/misc/ftml-test-cases-export.ftml:18]: Probably? It's probably not worth it hardcoding into the git attributes because the dev sandbox is an intentionally volatile module

File: client/modules/cm-lang-ftml/src/grammars/ftml.yaml (hunk 6 above)
Comment (Monkatraz) [reply client/modules/cm-lang-ftml/src/grammars/ftml.yaml:176]: true suffering

File: client/web/dev-sandbox/public/static/misc/ftml-test-cases-export.ftml (hunk 5 above)
Comment (emmiegit) [reply client/web/dev-sandbox/public/static/misc/ftml-test-cases-export.ftml:18]: It's volatile? What does this module do exactly?

File: client/modules/cm-lang-ftml/src/grammars/ftml.yaml (hunk 7 above)
Comment (Monkatraz) [reply client/modules/cm-lang-ftml/src/grammars/ftml.yaml:315]: I uh... don't know lmao

I disabled this rule because while it _usually_ works, it occasionally does weird screwy matches on stuff it shouldn't. I don't know why exactly this happens. I'll probably get back to this at some point and see if I can make a better rule for this.

File: client/web/dev-sandbox/public/static/misc/ftml-test-cases-export.ftml (hunk 5 above)
Comment (Monkatraz) [reply client/web/dev-sandbox/public/static/misc/ftml-test-cases-export.ftml:18]: It's literally a playground website I'm currently testing the editor on. It's in the repo because it does take a good bit of setup to use and it's generally useful for everyone to have access to. It also actually makes sure that the monorepo can build a website

File: client/modules/cm-lang-ftml/src/grammars/ftml.yaml (hunk 8 above)
Comment (Monkatraz) [reply client/modules/cm-lang-ftml/src/grammars/ftml.yaml:384]: oh yeah lmao this is the crazy regex for dumb links like `wikipedia:foo-bar` that wikidot supports

I'll add a comment

File: client/modules/cm-lang-ftml/src/grammars/ftml.yaml (hunk 9 above)
Comment (Monkatraz) [reply client/modules/cm-lang-ftml/src/grammars/ftml.yaml:613]: This is the rule for `[[css]]` blocks.

I made it a convention to name block containers that nest some other language `BlockNested`, mostly so that they're very obvious in the AST.

File: client/modules/cm-lang-ftml/src/grammars/ftml.yaml (hunk 10 above)
Comment (Monkatraz) [reply client/modules/cm-lang-ftml/src/grammars/ftml.yaml:999]: this is like my third attempt at making a grammar definition format

at some point I have to get it right lmao

File: client/modules/cm-lang-ftml/src/grammars/ftml.yaml (hunk 7 above)
Comment (emmiegit) [reply client/modules/cm-lang-ftml/src/grammars/ftml.yaml:315]: Okay, just add a `TODO` then I guess.

File: client/modules/cm-lang-ftml/src/grammars/ftml.yaml (hunk 8 above)
Comment (emmiegit) [reply client/modules/cm-lang-ftml/src/grammars/ftml.yaml:384]: Yeah... Real nonsense.

File: client/modules/cm-lang-ftml/src/grammars/ftml.yaml (hunk 9 above)
Comment (emmiegit) [reply client/modules/cm-lang-ftml/src/grammars/ftml.yaml:613]: Ah, okay.

File: client/modules/cm-lang-ftml/src/grammars/tokens.yaml (hunk 3 above)
Comment (Monkatraz) [reply client/modules/cm-lang-ftml/src/grammars/tokens.yaml:13]: Made [an issue](https://scuttle.atlassian.net/browse/WJ-834) for documentation in general.

**Comment by Monkatraz** [comment]
thanks @ammongit 

I made the requested changes in the followup PR I've been working on.
